### PR TITLE
[CONFIGURATION] File configuration - Tracer/Meter/Logger config

### LINKED ISSUE #3915: [CONFIGURATION] File configuration - Tracer/Meter/Logger config
Implement the following types from the declarative configuration spec:

* type: ExperimentalLoggerConfig
* type: ExperimentalLoggerConfigurator
* type: ExperimentalLoggerMatcherAndConfig
* type: ExperimentalMeterConfig
* type: ExperimentalMeterConfigurator
* type: ExperimentalMeterMatcherAndConfig
* type: ExperimentalTracerConfig
* type: ExperimentalTracerConfigurator
* type: ExperimentalTracerMatcherAndConfig

### PR DESCRIPTION
**Pull request by UMMAN2005**
Fixes #3915

## Summary

Implements the file configuration types for Tracer, Meter, and Logger as described in [#3915](https://github.com/open-telemetry/opentelemetry-cpp/issues/3915).

### Changes

- **9 new configuration model headers**: `ExperimentalTracerConfig`, `ExperimentalTracerMatcherAndConfig`, `ExperimentalTracerConfigurator` (and their Meter/Logger equivalents), following the project's one-class-per-header convention.
- **Updated provider configurations**: Replaced `FIXME` comments in `TracerProviderConfiguration`, `MeterProviderConfiguration`, and `LoggerProviderConfiguration` with the new configurator fields.
- **YAML parsing**: Added parsing support for the `tracer_configurator/development`, `meter_configurator/development`, and `logger_configurator/development` sections in `ConfigurationParser`.
- **SDK builder wiring**: Implemented `CreateTracerConfigurator`, `CreateMeterConfigurator`, `CreateLoggerConfigurator` in `SdkBuilder` to construct `ScopeConfigurator<T>` objects from parsed configuration, including wildcard name matching (`?` and `*`).
- **Unit tests**: Added tests for each signal's configurator parsing in `yaml_trace_test.cc`, `yaml_metrics_test.cc`, and `yaml_logs_test.cc`.

### Types implemented

```
ExperimentalTracerConfig
ExperimentalTracerConfigurator
ExperimentalTracerMatcherAndConfig
ExperimentalMeterConfig
ExperimentalMeterConfigurator
ExperimentalMeterMatcherAndConfig
ExperimentalLoggerConfig
ExperimentalLoggerConfigurator
ExperimentalLoggerMatcherAndConfig
```

## Test plan

- [x] Added unit tests for parsing configurator sections (no configurator, default only, default + matchers)
- [ ] CI builds and tests pass on all supported platforms

### CODE REVIEW

File: sdk/src/configuration/sdk_builder.cc [hunk 1]
```diff
@@ -984,6 +1042,35 @@ std::unique_ptr<opentelemetry::sdk::trace::SpanProcessor> SdkBuilder::CreateSpan
   return sdk;
 }
 
+std::unique_ptr<opentelemetry::sdk::instrumentationscope::ScopeConfigurator<
+    opentelemetry::sdk::trace::TracerConfig>>
+SdkBuilder::CreateTracerConfigurator(
```
Comment (lalitb) [inline sdk/src/configuration/sdk_builder.cc:1010]: `CreateTracerConfigurator`, `CreateMeterConfigurator`, and `CreateLoggerConfigurator` are nearly identical. Not a blocker, as this follows the existing code pattern we have, but good to see if we can use the template helper to reduce this duplication.

File: sdk/src/configuration/configuration_parser.cc [hunk 2]
```diff
@@ -1883,6 +1989,50 @@ std::unique_ptr<SpanProcessorConfiguration> ConfigurationParser::ParseSpanProces
...
+  {
+    for (auto it = child->begin(); it != child->end(); ++it)
+    {
+      std::unique_ptr<DocumentNode> element(*it);
+      model->tracers.push_back(ParseExperimentalTracerMatcherAndConfigConfiguration(element));
+    }
+  }
+
+  return model;
+}
```
Comment (lalitb) [inline sdk/src/configuration/configuration_parser.cc:2029]: These 3 methods are mostly identical with their metrics and logs equivalent. Good to see if we can reduce the duplication with the templatised versions.

File: sdk/src/configuration/sdk_builder.cc [hunk 3]
```diff
@@ -166,6 +184,46 @@ namespace sdk
 namespace configuration
 {
 
+static bool WildcardMatch(const std::string &pattern, const std::string &text)
```
Comment (lalitb) [inline sdk/src/configuration/sdk_builder.cc]: `WildcardMatch` is file-scoped static, making it impossible to directly unit-test or reuse. Please move it to a shared utility header (e.g., sdk/src/configuration/wildcard_match.h) and add dedicated unit tests covering the edge cases.

File: sdk/include/opentelemetry/sdk/configuration/meter_config_configuration.h [hunk 4]
```diff
@@ -0,0 +1,24 @@
...
+
+#include "opentelemetry/version.h"
+
+OPENTELEMETRY_BEGIN_NAMESPACE
+namespace sdk
+{
+namespace configuration
+{
+
+// YAML-SCHEMA: schema/meter_provider.yaml
```
Comment (lalitb) [inline sdk/include/opentelemetry/sdk/configuration/meter_config_configuration.h:14]: the meter config here says`// YAML-SCHEMA: schema/meter_provider.yaml` while tracer/logger headers say `.json`. Should be consistent across all three signals.

File: sdk/include/opentelemetry/sdk/configuration/configuration_parser.h [hunk 5]
```diff
@@ -164,6 +167,16 @@ class ConfigurationParser
   std::unique_ptr<LoggerProviderConfiguration> ParseLoggerProviderConfiguration(
       const std::unique_ptr<DocumentNode> &node) const;
 
+  ExperimentalLoggerConfigConfiguration ParseExperimentalLoggerConfigConfiguration(
+      const std::unique_ptr<DocumentNode> &node) const;
```
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/configuration/configuration_parser.h]: Some global naming comments, that apply everywhere:

The yaml schema type is ExperimentalFoo, but the corresponding C++ code should not say `Experimental` all the time.

Please change:

* the file names
* the classes names
* the method names

to remove the `Experimental` part.

File: sdk/include/opentelemetry/sdk/configuration/experimental_logger_config_configuration.h [hunk 6]
```diff
@@ -0,0 +1,24 @@
...
+{
+namespace configuration
+{
+
+// YAML-SCHEMA: schema/logger_provider.json
+// YAML-NODE: ExperimentalLoggerConfig
+class ExperimentalLoggerConfigConfiguration
+{
+public:
+  bool disabled{false};
```
Comment (marcalff) [inline sdk/include/opentelemetry/sdk/configuration/experimental_logger_config_configuration.h]: The yaml attribute has been renamed to `enabled` in the spec, in a recent change:

* https://github.com/open-telemetry/opentelemetry-configuration/pull/493

Rename to enabled.

File: sdk/include/opentelemetry/sdk/configuration/meter_config_configuration.h (hunk 4 above)
Comment (marcalff) [reply sdk/include/opentelemetry/sdk/configuration/meter_config_configuration.h:14]: It should be yaml everywhere.

File: sdk/src/configuration/sdk_builder.cc (hunk 1 above)
Comment (marcalff) [reply sdk/src/configuration/sdk_builder.cc:1010]: These are very similar because the configuration spec for each signal only contains a boolean so far, so they have the same structure.

Later, it is likely that configuration for each signal will diverge, for example adding a log level in the log config, or aggregation properties in the metric config, so then having separate code for each signal will make sense.

The current code looks ok to me, even if it appear repetitive.

Comment (marcalff) [reply sdk/src/configuration/sdk_builder.cc:1010]: Marking as resolved then, can be revisited later.

File: sdk/src/configuration/configuration_parser.cc (hunk 2 above)
Comment (marcalff) [reply sdk/src/configuration/configuration_parser.cc:2029]: Marking as resolved, can be revisited later.

**Comment by marcalff** [comment]
@UMMAN2005 

Thanks for the PR, it looks good with a few comments to fix.

Would you have time to address them ? Looking to merge this.

**Comment by UMMAN2005** [comment]
Hi, @marcalff 

I have updated code according to comments. 

Could you please check it out.

File: sdk/test/configuration/CMakeLists.txt [hunk 7]
```diff
@@ -3,6 +3,7 @@
 
 foreach(
   testname
+  wildcard_match_test
```
Comment (marcalff) [inline sdk/test/configuration/CMakeLists.txt]: This is a new test, it needs to be added in the bazel `BUILD` files as well.

File: sdk/src/common/wildcard_match.h [hunk 8]
```diff
@@ -0,0 +1,63 @@
...
+namespace sdk
+{
+namespace configuration
+{
+
+/**
+ * Matches a text string against a pattern containing '?' (any single character)
+ * and '*' (any sequence of characters, including empty) wildcards.
+ */
+inline bool WildcardMatch(const std::string &pattern, const std::string &text)
```
Comment (marcalff) [inline sdk/src/common/wildcard_match.h:21]: Since WildcardMath() is a general utility, one possibility is to:

* move it to sdk/src/common
* fix the CMake and bazel build there
*  move the associated unit test to sdk/test/common as well

**Comment by marcalff** [comment]
@UMMAN2005 

Thanks for all the CI fixes, the last change should do it for IWYU.
